worker/upload.go

```diff
@@ -197,11 +197,10 @@ type (
 		mu          sync.Mutex
```

Comment:
Since this is already quite the refactor I'd also try to get rid of the back-references between `upload`, `slabUpload` and `sectorUpload` as part of this PR.
Meaning that a `slabUpload` shouldn't require a `mgr` or `upload`, the `sectorUploadReq` shouldn't require a `upload` field etc.

e.g. when a new `slabUpload` is created, we can create it with a `candidates map[types.PublicKey]*uploader` field that is already populated.  These are then passed down to `sectorUpload` which has a `unusedCandidates map[types.PublicKey]` field and so on.

By passing all the required information down like that we shouldn't have a child type access a parent type and acquire the mutex which makes locking a lot safer.

It doesn't really make sense to have the `isUsed` method in the `upload` when trying to figure out whether a host has been used for a `slabUpload` upload for example. Which in turn means we don't need the `sid` because we can get rid of the `map[slabID]map[types.FileContracdtID]struct{}` map. Overall I think the code will get cleaner that way.

I'd also try to use `types.PublicKey` instead of `types.FileContractID` wherever possible. e.g. why check if 2 different contract IDs have been used if you can just check a single public key of the host? If we pass down the candidate `*uploader` to the `slabUpload`, it can implicitly handle a renewal without the `slabUpload` having to worry about it.